packages/core/src/utils/saml-assertion-handler.ts

```diff
@@ -0,0 +1,102 @@
+import type { ConnectorSession } from '@logto/connector-kit';
+import { connectorSessionGuard } from '@logto/connector-kit';
+import { errors } from 'oidc-provider';
+import type Provider from 'oidc-provider';
+import type { InteractionResults } from 'oidc-provider';
+import { z } from 'zod';
+
+import assertThat from './assert-that.js';
+
+/**
+ *
+ * The following three methods (`getInteractionFromProviderByJti`, `assignResultToInteraction`
+ * and `epochTime`) refer to implementation in
+ * https://github.com/panva/node-oidc-provider/blob/main/lib/provider.js
+ */
+type Interaction = Awaited<ReturnType<Provider['interactionDetails']>>;
+
+const epochTime = (date = Date.now()) => Math.floor(date / 1000);
+
+const getInteractionFromProviderByJti = async (
+  jti: string,
+  provider: Provider
+): Promise<Interaction> => {
+  const interaction = await provider.Interaction.find(jti);
+
+  if (!interaction) {
+    throw new errors.SessionNotFound('interaction session not found');
+  }
+
+  if (interaction.session?.uid) {
+    const session = await provider.Session.findByUid(interaction.session.uid);
+
+    if (!session) {
+      throw new errors.SessionNotFound('session not found');
+    }
+
+    if (interaction.session.accountId !== session.accountId) {
+      throw new errors.SessionNotFound('session principal changed');
+    }
+  }
+
+  return interaction;
+};
+
+const assignResultToInteraction = async (interaction: Interaction, result: InteractionResults) => {
```

Comment:
nit: could move interaction-specific methods to interaction utils, maybe, just for future use.

Reply:
Updated.